### PR TITLE
[22.01] Create config package, move app-related functionality to app

### PR DESCRIPTION
**Pull request by mvdbeek**
Work towards https://github.com/galaxyproject/galaxy/issues/13451

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

**Comment by nsoranzo** [comment]
Does this need to target 22.01? Seems quite "invasive" so late in the release cycle.

**Comment by mvdbeek** [comment]
I think so, otherwise we have to hack around the path resolution to the interactivetools_map setting.

**Comment by mvdbeek** [comment]
Though we could publish the config package from the dev branch and depend on that in gravity.

**Comment by mvdbeek** [comment]
(It also just moves around a few things around, it doesn't add new code, I don't think the lines changed number gives a good impression of the changes)

**Comment by nsoranzo** [comment]
Fine by me, #13451 didn't directly mention issues with interactivetools_map , good to have it stated somewhere for future reference!

File: test/unit/conftest.py
Comment (nsoranzo) [inline test/unit/conftest.py]: I think this fixture is not automatically applied to unit tests in packages?

Comment (mvdbeek) [reply test/unit/conftest.py]: Probably not, yeah. I think I'll try just setting `override_tempdir=False` in the relevant places

**Comment by nsoranzo** [comment]
A few scripts (`scripts/cleanup_datasets/pgcleanup.py`, `scripts/runtime_stats.py` and `scripts/set_user_disk_usage.py`) needs to updated to import `get_database_url()` from the right place.

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]:
```suggestion
```

File: lib/galaxy/config/__init__.py
Comment (nsoranzo) [inline lib/galaxy/config/__init__.py]:
```suggestion
                conf['filename'] = conf.pop('filename_template').format(**facts)
```

File: packages/config/HISTORY.rst
Comment (nsoranzo) [inline packages/config/HISTORY.rst]:
```suggestion

* Initial release
```

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]: Move to top?

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]:
```suggestion
```

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]:
```suggestion
        self.datatypes_registry = Registry(self.config)
```

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]: Move to top?

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]:
```suggestion
        self.security = IdEncodingHelper(id_secret=self.config.id_secret)
```

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]: Move to top?

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]: Move to top?

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]: Move to top?

File: lib/galaxy/app.py
Comment (nsoranzo) [inline lib/galaxy/app.py]: Move to top?

File: packages/config/LICENSE
Comment (nsoranzo) [inline packages/config/LICENSE]: We can't change the license like that for the files which are already released under the AFL-3.0. I think this should symlink to `../../LICENSE.txt` for the time being.

File: packages/config/galaxy/__init__.py
Comment (nsoranzo) [inline packages/config/galaxy/__init__.py]: Files in `packages` are not formatted with `make format`, so better to get it right before merging forward.

```suggestion
__path__ = __import__("pkgutil").extend_path(__path__, __name__)  # type: ignore[has-type]
```

File: packages/app/requirements.txt
Comment (nsoranzo) [inline packages/app/requirements.txt:5]:
```suggestion
galaxy-job-metrics
galaxy-objectstore
```

File: packages/config/galaxy/project_galaxy_util.py
Comment (nsoranzo) [inline packages/config/galaxy/project_galaxy_util.py]:
```suggestion
PROJECT_AUTHOR = "Galaxy Project and Community"
PROJECT_DESCRIPTION = "Galaxy Configuration"
PROJECT_EMAIL = "galaxy-committers@lists.galaxyproject.org"
RAW_CONTENT_URL = "https://raw.github.com/{}/{}/master/".format(PROJECT_USERAME, PROJECT_NAME)
```

File: packages/config/test-requirements.txt
Comment (nsoranzo) [inline packages/config/test-requirements.txt]: Not sure if that's needed.

```suggestion
```

File: packages/config/tests/util
Comment (nsoranzo) [inline packages/config/tests/util]: Drop this symlink?

File: packages/util/galaxy/version.py
Comment (nsoranzo) [inline packages/util/galaxy/version.py]: The `packages/app/galaxy/version.py` symlink (that also points to this file) needs to be removed.

**Comment by ic4f** [comment]
> I think so, otherwise we have to hack around the path resolution to the interactivetools_map setting.

If we merge this into 22.01, I'll have to hack around the alembic branch. A lot. Why 22.01? This is a considerable change to the code base (most welcomed and needed, but considerable).

**Comment by mvdbeek** [comment]
> If we merge this into 22.01, I'll have to hack around the alembic branch. A lot. Why 22.01? This is a considerable change to the code base (most welcomed and needed, but considerable).

We can't support interactivetools with gravity otherwise.

File: packages/config/LICENSE
Comment (mvdbeek) [reply packages/config/LICENSE]: I think the contributions are all covered by the relicensing clause we tick for new PRs.

**Comment by ic4f** [comment]
> > If we merge this into 22.01, I'll have to hack around the alembic branch. A lot. Why 22.01? This is a considerable change to the code base (most welcomed and needed, but considerable).
> 
> We can't support interactivetools with gravity otherwise.

So the reason for creating a config package in this PR is to fix interactivetools in 22.01 which was broken with the introduction of gravity? And that's the only possible fix? 

If so, here's what this will overwrite (at least  - that's just what immediately comes to mind)
https://github.com/galaxyproject/galaxy/pull/13108/files#diff-d893d6e637fce7baa9553158dd9ba652cf39a80f4b293a8a07a457747955db85

https://github.com/galaxyproject/galaxy/pull/13108/files#diff-dc748c7359c499f7b942f1ea1736043676e1fba44ab06c0b157108f1155f3e7a

**Comment by mvdbeek** [comment]
I can rebase your PR, this PR here is a minor change.

File: lib/galaxy/config/__init__.py
Comment (jdavcs) [inline lib/galaxy/config/__init__.py]: If we can't/shouldn't get here, the method should be removed (otherwise it's more dead code). It's test-related; introduced here: https://github.com/galaxyproject/galaxy/commit/f34bb5e1f66e47ae13396f8ab8d8336aa635a255 , then modified (see discussion) here: https://github.com/galaxyproject/galaxy/pull/9631#issuecomment-617008412

Comment (mvdbeek) [reply lib/galaxy/config/__init__.py]: Debugging leftover, thanks for the catch.

**Comment by ic4f** [comment]
> I can rebase your PR, this PR here is a minor change.

xref from committers channel: Holding fingers crossed for minor change, will handle rebase myself. Thank you for addressing this!

File: packages/config/requirements.txt
Comment (nsoranzo) [inline packages/config/requirements.txt:3]:
```suggestion
boltons
pykwalify
PyYAML
requests
```

**Comment by jmchilton** [comment]
> Though we could publish the config package from the dev branch and depend on that in gravity.

Galaxy depends on gravity. I don't think you can place Galaxy packages in Galaxy's virtualenv if you're running Galaxy from the root tree. This is why we have pulsar-galaxy-lib and don't just depend on galaxy-app. Even if it did work (and I think it won't) which package would you want to be the source of truth for an installed wheel or the root? Or is the idea to have a duplicated package without the dependency like pulsar-galaxy-lib?

**Review by nsoranzo — COMMENTED** [review]
Can you please move `pykwalify` from `packages/app/requirements.txt` to `packages/app/test-requirements.txt` ?

**Comment by mvdbeek** [comment]
> Galaxy depends on gravity. I don't think you can place Galaxy packages in Galaxy's virtualenv if you're running Galaxy from the root tree.

That's a good point, but I'm pretty sure that works and gravity would use the virtualenv package, while a root install uses local imports (I think that's because we set PYTHONPATH. Or so this seems to be the case for planemo installed in a Galaxy virtualenv.

I tested this real quick now by editing version.py to 10.01 in the galaxy-util package, then stashing the changes:
```console
python -c "from galaxy import version; print(version.VERSION)"
10.01.rc1
PYTHONPATH=lib/ python -c "from galaxy import version; print(version.VERSION)"
22.01.rc1
```
Do you think this is good enough, or do you think we need to add a workaround like `pulsar-galaxy-lib` ?

**Comment by mvdbeek** [comment]
> Can you please move `pykwalify` from `packages/app/requirements.txt` to `packages/app/test-requirements.txt` ?

Do we need it at all in packages/app, given that app depends on config ?

File: packages/config/requirements.txt
Comment (mvdbeek) [reply packages/config/requirements.txt:3]: These are transitive dependencies, coming in via galaxy-util. Is our goal to list all of these ?

**Comment by nsoranzo** [comment]
> > Can you please move `pykwalify` from `packages/app/requirements.txt` to `packages/app/test-requirements.txt` ?
> 
> Do we need it at all in packages/app, given that app depends on config ?

Following "explicit is better than implicit", I prefer we add all packages that are imported in some package B files even if they are already dependency of a requirement A. In this way, if the package A drops the dependency at some point, package B doesn't break.

**Comment by mvdbeek** [comment]
Uff, that seems hard to maintain as a policy without tooling to enforce this. Are we sure we're doing that now ?

**Comment by nsoranzo** [comment]
> Uff, that seems hard to maintain as a policy without tooling to enforce this. Are we sure we're doing that now ?

As a best effort. We also sometimes miss requirements in `pyproject.toml` that gets installed as deps of deps.

**Comment by jmchilton** [comment]
Planemo should never be installed in Galaxy's virtualenv 😅 - this used to crash and burn terribly and that is what I would expect. This seems problematic to me, Python isn't node.

I don't think we should have two sources of truth in Galaxy's virtualenv. I don't think this makes sense for a production setting. Bringing gravity into Galaxy's package system isn't a valid way to deal with this?

**Comment by mvdbeek** [comment]
> this used to crash and burn terribly and that is what I would expect.

I think that wasn't necessarily due to fundamental issues with the approach, but how we propagated the VIRTUAL_ENV variable. But of course it's not a good idea since the non-galaxy dependencies could conflict.

> Bringing gravity into Galaxy's package system isn't a valid way to deal with this?

We could of course, but how would this work ? I'd assume we'd still have to install the gravity package ?

File: packages/config/galaxy/project_galaxy_util.py
Comment (nsoranzo) [inline packages/config/galaxy/project_galaxy_util.py]: This file should be renamed `project_galaxy_config.py` .

**Comment by nsoranzo** [comment]
> Bringing gravity into Galaxy's package system isn't a valid way to deal with this?

I did propose something along these lines: https://github.com/galaxyproject/galaxy/pull/13224#issuecomment-1018859374 but @natefoo was not in favour.

**Comment by mvdbeek** [comment]
I'd also assume that production setups wouldn't necessarily install gravity into the same virtualenv as Galaxy, we only do that for simplicity in the run.sh case. Maybe we should drop gravity from `requirements.txt` and install it somewhere in common_startup.sh ?

**Review by nsoranzo — APPROVED** [review]

**Comment by nsoranzo** [comment]
> I'd also assume that production setups wouldn't necessarily install gravity into the same virtualenv as Galaxy, we only do that for simplicity in the run.sh case. Maybe we should drop gravity from `requirements.txt` and install it somewhere in common_startup.sh ?

Maybe we can install gravity with [pipx](https://pypa.github.io/pipx/) ?

**Comment by mvdbeek** [comment]
Maybe we could ignore the state dir issue for now. The documentation says to activate galaxy's virtualenv, which has a default state dir set. So this would still work, as long as Galaxy's virtualenv doesn't include a conflicting gravity. It does take away from the advantage of using pipx, but it's better than the other possible options at this point.

**Comment by nsoranzo** [comment]
@mvdbeek I am a bit lost with the plan, and I think we need buy-in from at least @natefoo and @jmchilton . Should we discuss this at the next backend-wg meeting on 2022-03-08, but merge this in the mean time?

**Comment by jmchilton** [comment]
I continue to hate this but I don't understand the underlying issues. We won't really have a very clean way to remove those older Galaxy packages out of the virtualenv if we decide to undo this in a future release. This isn't a -1 though, I hope I prove to be wrong about this - and this is just a once bitten, twice shy kind of response.

**Comment by jmchilton** [comment]
Also... to be clear... I do love having a standalone galaxy-config package. That is absolutely awesome regardless of the dependency issues with gravity.

**Pull request merged**
